### ISSUE TITLE
[css-scroll-snap-2] Better snap physics and customization?

### ISSUE THREAD

**Issue by flackr** [w3c/csswg-drafts#8549]
We've had requests from developers (e.g. [airbnb](https://bugs.chromium.org/p/chromium/issues/detail?id=1189696#c3), [google search](https://bugs.chromium.org/p/chromium/issues/detail?id=1420772)) to customize the fling curve of snap point scrolls. A common cause or desire is to align behavior, which itself seems partially motivated from the way that browsers haven't given special treatment to scrolling physics for certain snap use cases in the same way that the experiences being emulating have given them different physics.

`scroll-snap-stop` is often used as a simple form of control over the physics (forcing a stop at each item), but even when the typical interaction is to scroll one item at a time developers often want to still allow flings to cross multiple items.

I think expectations may vary based on the size of snap items and lengths of documents, e.g.:
* For long articles with headers that may want to snap, having snap behave similar to normal scrolling likely matches expectations
* For lists with snap points on each item, there is often an expectation that a standard length flick will move about a page of items. The experiment in [crbug.com/1189696](https://bugs.chromium.org/p/chromium/issues/detail?id=1189696) shows that for such use cases it may be as simple as adjusting fling distances to meet expectations here. A paginated book would likely have similar expectations.
* Scrolling within a large snap point should probably match document scrolling but may want to prefer seeing the end of the current snap point rather than jumping to the next snap point.

While some cases the browser may be able to "just do the right thing", I think there are cases that would be helpful if the developer could give the browser a hint what kind of category things fall into.

Some possible ideas:
* Add a `scroll-snap-distance: item | page | document` which hints whether the UA should treat a typical fling as traversing the distance of one item, one page, or a document scroll (the current behavior). Alternately, it could take a length.
* Add a `scroll-snap-friction: #` which applies some amount of friction to getting to scrolling past each subsequent snap point.
* Something else?

I'm not sure what the exact right solution is here but there seems to be a common developer need to apply different physics to different scenarios.

**Comment by tabatkins** [comment]
I'm pretty strongly against more arbitrary control over fling curves; every single site that does customized scrolling that doesn't match the my OS's default fling physics is deeply unsettling to interact with, personally. This probably applies to things like letting people choose a friction amount, especially since there's no intuition for how it works so it'll just be a random number that feels right to the dev at the time, and totally inconsistent across sites.

(In particular, I consider the [Google Search request](https://bugs.chromium.org/p/chromium/issues/detail?id=1420772) to be an **active anti-goal**. "Consistent experience across OSes and devices" is another way of saying "wrong physics on every device and OS (*maybe* matching the one OS/device the dev happens to use and like)". Users want physics to be consistent across different pages/apps on their device, because it's something you build muscle memory for; a single page being different from other pages, but consistent across different devices, has literally zero benefit for users and significant downsides, imo.)

But having a little more control over how far a fling is meant to go over stops sounds reasonable! Especially with nice semantic names that appear to be meaningful, like the item/page/document you have here. 

Is this meant to interact with scroll-snap-stop? Like, `item` would be the default and give the current stop `scroll-snap-stop:always` behavior, while the other values loosen it a bit? Or were you thinking about something else?

**Comment by flackr** [comment]
I guess it depends what you mean by loosen it up a bit. I was thinking that it would be closer to the experiment I did in [crbug.com/1189696](https://bugs.chromium.org/p/chromium/issues/detail?id=1189696) where the browser can determine more appropriate fling distances. The goal there was to make the average fling distance about a page but if the user flings fast they can still fling further. I was initially thinking this should just be the default but have since come to realize this behavior only makes sense for itemized list scrollers and not something like snap points added to spec sections.

That said, my exact proposal of changing the "average" (specific fling modification to be determined by the browser) distance would mean that you'd either use this new property (which is like a less strict `scroll-snap-stop`) or `scroll-snap-stop` but it would be interesting to think if there are ways to make the two properties work together in meaningful ways. E.g. maybe if this is like a stop at the next mandatory item after x pages of scrolling then we could do the velocity modifications in chrome similar to the [crbug.com/1189696](https://bugs.chromium.org/p/chromium/issues/detail?id=1189696) when you're in such a scroller because we know that you only meaningfully need to differentiate between the next N snap points and not scroll further.

**Comment by tabatkins** [comment]
Yeah, "add friction but allow it to be overcome" should, I think, be done as part of scroll-snap-stop, which is basically "infinite friction on each item". Relaxing that to be "large but not infinite friction" sounds like a useful thing in the continuum of this property.

Maybe two axises - how often we want the "large friction" bump (every item, every page, none) and then how large the friction should be (infinite, or just large but overcomeable with a large fling).

The current `always` value is "every item, infinite", but "every page, infinite" sounds reasonable, as do both "every item, large" and "every page, large".

**Comment by flackr** [comment]
So in terms of an algorithm the UA could apply, we could scan through snap points in the scroll direction and reduce fling distance in some UA-chosen way based on this value on encountering each scroll-snap-stop point.

I am a bit worried that if we don't apply the friction uniformly it will be really hard to scroll less than that distance. For example, imagine a scroller where 5 items fit per screen. If we apply more friction to every 5th item, then it will be really hard to scroll less than 5 items. This might be fine, but it feels like it would be better to express this as those other items not having any scroll-snap-stop power. Naively, I'd want the scroll-snap-stop friction of each point to add up such that it scrolls about a page worth of items for some standard length fling.

So TLDR, I'd propose the friction of each `scroll-snap-stop` item be based on its proportion of the developer specified target distance.

**Comment by tabatkins** [comment]
Why would it be hard to scroll less than 5? What would it be different vs scrolling without any added friction at all?

I imagine it might be more difficult to scroll, say, *6* items (because you'd have to hit the momentum *just right*) but I think that's okay?

**Comment by flackr** [comment]
For illustrative purposes let's say each item is 200px on a 1000px wide scroller, and that the increased scroll friction reduces the fling distance by 4000px.

This means that flinging anywhere from 1000px to 5000px will land you on item 5, and to get less than item 5 you'd need to fling less than 1000px.

My alternative proposal was that having `scroll-snap-stop-friction: page` would result in each item applying let's say 800px of fling reduction (e.g. snap stop width / page width * 4000) so scrolling 0px - 1000px lands on item 1, 1001px - 2000px lands on item 2, 2001px - 3000px lands on item 3, etc.

**Comment by tabatkins** [comment]
> to get less than item 5 you'd need to fling less than 1000px.

Right, but that's exactly what you'd require if we weren't applying any special friction at all, too. This is why I'm confused you're saying this is a problem caused by the friction.

**Comment by flackr** [comment]
The difference is that we've made a large range of flings scroll exactly 5 items. So scrolling less requires a very gentle fling compared to that which gets you into the large stop at 5. This issue came up because on horizontal mandatory snap scrollers users were flinging way too far with relatively short flings. If every item has an equivalent `scroll-snap-stop` applied I think it would match author intentions better to make them each have a similarly large fling range that stops on those items (i.e. effectively reducing the fling distance like my prototype patch).

**Comment by tabatkins** [comment]
Okay, so with the assumption that normal fling velocities are tuned for long documents rather than "a page", I can see why it's reasonable to essentially increase the friction on each item so that a standard fling will advance a page worth (rather than multiple pages, as normal), and then it's easier to fling less and get individual items aligned.

**Comment by argyleink** [comment]
Another factor to think about is how the scroller is only allowed to have [`mandatory || proximity`](https://developer.mozilla.org/en-US/docs/Web/CSS/scroll-snap-type) snap points. This would be like flexbox not having `align-self` or `place-self`. 

I'm wondering if something like `scroll-snap-point-type: mandatory || proximity`, set on an item basis, would help single scrollers be more flexible. This would help snap points be more viable on a full document, for example, as it's very annoying if the scroller is set to mandatory, the user can't place the scroller in any position they want. It would be nice if certain large landmarks could individually be mandatory, making them a bit more greedy in terms of proximity and magnetizing the scroller to it while all the other snap points are light suggestions.

**Comment by flackr** [comment]
I was originally confused by this - because if there exist mandatory scroll snap points then presumably you have to choose one of them to snap to. I think what you're asking for is similar to increasing the proximity snap distance or an alternative mandatory snap that only looks at on screen points so that they behave like mandatory snap points while in view but otherwise do not force the scroll back to them. You can do this by making the entire non-mandatory region of the scroll a big mandatory snap point itself but I can see how that's a bit awkward and doing so doesn't allow you to further place proximity snap points within that giant mandatory snap area.

**Comment by tabatkins** [comment]
Right, "mandatory" really means mandatory, and that only makes sense (under the current definitions) as a policy on the container. We could have had a different conception of mandatory that meant something like "this snap point must be snapped to if it's the nearest point to where the fling lands, regardless of the actual distance", but that would have been a different model.  But what Adam is describing actually just sounds like "stronger proximity", which is definitely workable in the model, it just wasn't pursued for v1.

**Comment by kcrwfrd** [comment]
Related ticket for iOS Safari: https://bugs.webkit.org/show_bug.cgi?id=243582

Essentially, the momentum scrolling behavior from a fast "flick" on iOS Safari is disabled when scroll snap is enabled.

I'm not sure if this is technically a "bug" or simply a lack of specification.

Desktop Chrome on macOS with a touch device like a touch pad or Magic Mouse behaves as desired, you can flick to quickly scroll to the end of the container.

Edit: interestingly, desktop Safari 18.6 with a touch device behaves as desired, too.

**Comment by flackr** [comment]
I have worked up a concrete proposal here, including more recent requests I've heard for resistance to leaving snap areas and difficulties I encountered with building pull to refresh with scroll snap ([demo](https://flackr.github.io/web-demos/css-scroll-snap/email/)). See motivating use cases for a rough representative set of things that are difficult to replicate today.

## Proposal
I think that we can address all of the motivating use cases that have been raised (see examples below) with a couple additional values and an additional property:
1. Resistance to bypassing snap points in flings - while flinging, the snap point will actively resist passing it (i.e. it will slow down your scroll).

```css
scroll-snap-stop: normal | always | <length> | before
```

`before`: No fling strength will reach this snap point, the user must drag at least as far as the point at which this snap point is preferred.
`<length>`: Some amount of fling strength will be "consumed" by bypassing this snap point. Always is effectively like specifying an infinite length.

2. Strength and magnetism to the snap point. If attract, applies magnetic / gravitational attraction to the scroll while actively dragging (e.g. notification dismissing behavior on android).

```css
scroll-snap-strength: [auto | <length>] || [attract | no-attract]
```

<length> is used to determine the relative attraction of this scroll snap area compared to snapping to a different snap point. If the lengths overlap, the inflection point where you switch from one to the other is determined by the average position between the maximum strengths of each group of aligned points, capped to the snap points themselves. If the scroll container has scroll-snap-type mandatory, then the midpoint between any remaining gap in lengths is used as the point at which you select one or the other.

If attract is specified, the scroll will resist scrolling away or accelerate to scrolling to the next point with attract specified. Once you exceed this length the UA may catch the scroll up, but I don't think this needs to be mandatory.

Note that we could consider having a small set of keywords instead of the `<length>` options if we think that arbitrary lengths may be too permissive. E.g. `strong`.

## Motivating use cases

### Pull to refresh
A snap point that you have to scroll all the way to, and cannot fling to.
```css
.main { scroll-snap-align: top; }
.ptr { scroll-snap-align: top; scroll-snap-stop: before; scroll-snap-strength: 0; }
```

### Dismissable areas (e.g. bottom sheet)
A snap point that resists actively scrolling it away.
```css
.sheet {
  scroll-snap-strength: 3rem attract;
}
```

### Galleries or carousels
Increased resistance to flinging past many entries - like a weaker scroll-snap-stop.
```css
.item {
  scroll-snap-align: center;
  scroll-snap-stop: 30vw;
}
```

**Comment by tabatkins** [comment]
To be clear: `scroll-snap-strength` just controls the attraction strength of `proximity` stops, right? Mandatory stops are, effectively, already `calc(infinity) attract`, yeah?

How do these interact with directional scrolls like arrow keys? Particularly, how `before` and `attract` work. These both seem to be defined to give convenient, useful behaviors for flinging/dragging, but I feel like neither are capable of having any effect on arrow keys. Is the "just ignore them" behavior what we want?

**Comment by nickcoury** [comment]
I love the intent of this, but don't love the API or concrete proposal. I agree with the first comment in the thread that half the reason it's a problem is that the behavior on mobile doesn't match the underlying native platform that users develop a lot of muscle memory for. The mechanisms here seem to be providing configuration for yet another physics system that doesn't match the native platforms.

In particular, Flutter has useful prior art when they tried to make a unified API and received significant pushback on both Android and iOS for feeling non-native which led them to completely change their scroll snapping logic to more closely resemble native platforms:

https://github.com/flutter/flutter/issues/119875
https://github.com/flutter/flutter/issues/149221
https://github.com/flutter/flutter/issues/120338

I think this API makes a bit more sense for use cases that aren't heavily scrolling based, like pull to refresh. But will be problematic for heavy scrolling cases like carousels. I'm not even sure how I'd tune this for a good experience. It would be trial and error and very inconsistent for users across sites.


I’ve been digging into other implementations (Android, iOS, Flutter, etc) to see if that suggests a better balance between platform-specific behavior and customization, and it hints at some other possible designs. In particular, there seem to be two conceptual phases:

Phase 1 - Fling (Momentum Decay). The OS uses standard, unmodified platform friction to let the swipe coast and mathematically project a theoretical resting coordinate. Modifying friction here is what feels uncanny valley.

Phase 2 - Settle (Snap). Once a snap point is determined, the OS abandons the friction decay and transitions to a spring curve using the active velocity as an input.

Here are some APIs that might map to that. Please forgive any errors in use of native APIs, I don't have deep enough experience with them and this was based on quick research.

### 1. scroll-snap-limit
Give web authors a mechanism to cap how far the fling can go, but still let the OS control the fling.

```css
.carousel {
  /* Flings as far as native momentum permits */
  scroll-snap-limit: none;
  
  /* Fling traverses a maximum of 1 item */
  scroll-snap-limit: 1;
  
  /* Fling traverses a maximum of 1 visible viewport */
  scroll-snap-limit: 100vw;
}
```

Both Android and iOS use this concept:

iOS
```Swift
ScrollView { ... }
  .scrollTargetBehavior(.viewAligned(limitBehavior: .alwaysByOne))
```
Android
```Kotlin
val flingBehavior = PagerDefaults.flingBehavior(
    pagerSnapDistance = PagerSnapDistance.atMost(1)
)
```

### 2. scroll-fling-decay
I'm not as confident on this one, and I'm more in favor of skipping it. But if modifying fling friction is required, there's a bit of prior art on the native side.

```css
.carousel {
  /* Uses standard OS momentum */
  scroll-fling-decay: auto; 
  
  /* Maps directly to steeper native OS deceleration */
  scroll-fling-decay: fast; 
}
```

Native equivalents:

iOS
```Swift
// Note this is in UIKit which is legacy, and there's not an equivalent option in SwiftUI
scrollView.decelerationRate = .fast 
```
Android
```Kotlin 
val decay = exponentialDecay(frictionMultiplier = 2.0f)
```

### 3. scroll-snap-animation
For the second phase snap, developers may want to modify how fast it snaps which gives most of the feel that's worth customizing anyways.

Most systems now use spring physics which feel intuitive and natural, and can readily take the current velocity as an input. While this means spring() would likely need to be finalized as part of this, that may be a desirable thing anyways. There's an option here to use one or both of:
1. Presets that map to the OS or are tuned by the UA to feel natural.
2. expose a more raw spring() configuration object that uses the velocity. The values likely need to be clamped to avoid situations like infinite overbouncing.

```css
.carousel {
  /* Semantic OS defaults */
  scroll-snap-animation: auto | fast | smooth | bouncy; 
  
  /* Customize */
  scroll-snap-animation: spring(<mass> <stiffness> <damping>);
}
```

Native equivalents:

Android
```Kotlin
val flingBehavior = snapFlingBehavior(
    snapAnimationSpec = spring(
        stiffness = Spring.StiffnessLow, 
        dampingRatio = Spring.DampingRatioMediumBouncy
    ) 
)
```

iOS doesn't actually have an equivalent here, except maybe a custom scroll implementation in UIKit that isn't in SwiftUI. It sounds like they don't allow this kind of scroll snap customization on native scrollers. Programmatic scrolls can be customized, and that's still done with springs.
```Swift
ScrollView { ... }
  .scrollTargetBehavior(.viewAligned)
  // This line only affects programmatic scrolls
  .animation(.spring(response: 0.3, dampingFraction: 0.6)) 
```

### Non-carousel use cases
I think these options would still largely meet the intent there. Bottom sheets could still be heavily tweaked by shifting snap points and using these options. Pull to refresh is essentially all the settling phase with no fling velocity.

**Comment by flackr** [comment]
> The mechanisms here seem to be providing configuration for yet another physics system that doesn't match the native platforms.

The options here are tweaks on top of native platform scrolling behavior intended to allow developers to provide hints for what to snap to. It is not prescribing or changing anything about what the physics system is, as that is intentionally platform specific and left to the UA to correctly implement the platform phsyics. The idea is that the UA will use these hints to provide 

> I’ve been digging into other implementations (Android, iOS, Flutter, etc) to see if that suggests a better balance between platform-specific behavior and customization, and it hints at some other possible designs. In particular, there seem to be two conceptual phases:
> 
> Phase 1 - Fling (Momentum Decay). The OS uses standard, unmodified platform friction to let the swipe coast and mathematically project a theoretical resting coordinate. Modifying friction here is what feels uncanny valley.
> 
> Phase 2 - Settle (Snap). Once a snap point is determined, the OS abandons the friction decay and transitions to a spring curve using the active velocity as an input.

I'm not sure what the difference or the point in time at which you switch from phase 1 to phase 2 is, I have not seen any platforms where you can see a change in the fling from momentum to "settling". In particular if you fling in the OS application switcher on Android or iOS there seems to be no transition point from momentum decay, it appears to simply decay smoothly in a way that it lands on a snap point.

> Here are some APIs that might map to that.

How to scroll to a snap point once you've selected one is outside the scope of this proposal. The properties proposed here affect how you select a target snap point, but once you start a fling it is still up to the UA to provide a native-like scroll to the target.

While customizing this is interesting, I think such control would be orthogonal to the controls here which aim to provide a native-like scroll with some customization of how far you can fling.

> Please forgive any errors in use of native APIs, I don't have deep enough experience with them and this was based on quick research.
> 
> ### 1. scroll-snap-limit
> Give web authors a mechanism to cap how far the fling can go, but still let the OS control the fling.
> 
> .carousel {
>   /* Flings as far as native momentum permits */
>   scroll-snap-limit: none;
>   /* Fling traverses a maximum of 1 item */
>   scroll-snap-limit: 1;

We already have scroll-snap-stop: always to progress one snap point at a time, this was a natural property to extend to support more customization as has been proposed by the resolution here. Right now we have proposed distance based consumption of the fling distance. The problem I see with adding a number is that it necessarily has to be a property of the scroller rather than the snap point since it would be nonsensical to have two snap points with different number limits. We could consider this as an additional orthogonal property.

>   /* Fling traverses a maximum of 1 visible viewport */
>   scroll-snap-limit: 100vw;

I'm concerned that different scrolling methods may have different inherent momentums and so setting a single value rather than a reduction of the momentum may result in broken experiences. E.g. touchpad scrolling is often inherently faster than touch scrolling, forcing developers to set a single value may result in an experience which feels uncanny on the device they did not test while designing it.

> }
> Both Android and iOS use this concept:
> 
> iOS
> 
> ScrollView { ... }
>   .scrollTargetBehavior(.viewAligned(limitBehavior: .alwaysByOne))
> Android
> 
> val flingBehavior = PagerDefaults.flingBehavior(
>     pagerSnapDistance = PagerSnapDistance.atMost(1)
> )

iOS has [three values for limitBehavior](https://developer.apple.com/documentation/swiftui/viewalignedscrolltargetbehavior/limitbehavior), always stop (which we have with scroll-snap-stop: always), never (default in css), and few which doesn't let you decide precisely but reduces the native momentum. This suggests that the limiting behavior is really a reduction in system momentum rather than prescribing an absolute limit. This also seems to be what safari is doing in horizontal scroll snap areas by default today. This could be done via the resolved properties as `scroll-snap-stop: strong` or some non-zero length.

> ### 2. scroll-fling-decay
> ### 3. scroll-snap-animation

This seems out of scope for the problem of determining where you will snap to, right? I.e. let's open up a separate issue for this. I do worry this would mean the browser is not allowed to match the platform's physics, which is a concern you raised at the start.

**Comment by nickcoury** [comment]
I apologize, I missed the mark especially after reading your responses.

I'll explain my thinking better below. My overall biggest discussion point is making sure this feels consistent with the native OS scroll snapping on each platform so users can lean into muscle memory, and making sure there's not a foot gun to make it easily feel "off". If it works with that then that was my primary concern.

> It is not prescribing or changing anything about what the physics system is, as that is intentionally platform specific and left to the UA to correctly implement the platform physics.

I may have been thrown off by "magnetic / gravitational attraction" as I read that as proposing the math to use and it sounded like new physics. If I read this as more illustrative and it's still up to the UA to provide the physics, then it's more in line with what I imagined.

And if the UA is using native physics, I'll actually retract my suggestions for more customization. I was suggesting customizing spring decay as an alternative to magnetic/gravitational, but ideally if the platforms are dictating the most appropriate curve and we are hitting the needed use cases, it's probably better to avoid additional customizations. Otherwise I'm essentially suggesting the foot guns I wanted to avoid.

> the snap point will actively resist passing it (i.e. it will slow down your scroll)

This is the other part that I may need help understanding. I visualized this as rolling something near magnets, where they would slow down when near each snap point but not necessarily in between snap points when going past multiple. If it doesn't work that way and still feels true to existing patterns, then I'll assume I just misunderstood.

> no transition point

This is expected and works because of the velocity handoff makes it completely seamless. If I'm following the code right, Android handles the fling using `decayAnimationSpec` which is a spline model, then hands off to spring settling logic [here](https://cs.android.com/androidx/platform/frameworks/support/+/androidx-main:compose/foundation/foundation/src/commonMain/kotlin/androidx/compose/foundation/gestures/snapping/SnapFlingBehavior.kt;l=150-158;drc=e57dfa5d3c2605c905596aeb3e60ee9da66f9002) where `snapAnimationSpec` is the spring animation. Specs defined [here](https://cs.android.com/androidx/platform/frameworks/support/+/androidx-main:compose/foundation/foundation/src/commonMain/kotlin/androidx/compose/foundation/gestures/snapping/SnapFlingBehavior.kt;l=237-238;drc=e57dfa5d3c2605c905596aeb3e60ee9da66f9002)

Apple is a bit harder as their [WebKit Implementation](https://github.com/WebKit/WebKit/blob/cb78589da1c78d37b37c6f118003c959f0e81002/Source/WebCore/platform/ScrollingMomentumCalculator.cpp#L208) is strictly exponential decay, but that seems to be a fallback that's overridden by native code. I can find several examples of overshooting/rubber banding in iOS (home screen, AirBnB image carousels) which wouldn't happen with exponential decay so I believe this strongly indicates it's using springs to settle the curve. Apple has talked heavily about this in the past though, especially when they overhauled their physics system with the iPhone X e.g. [this talks about exponential decay](https://medium.com/@esskeetit/scrolling-mechanics-of-uiscrollview-142adee1142c) and the video it links talks about their move to spring physics in the middle.

> We already have scroll-snap-stop

Makes sense!

> different scrolling methods may have different inherent momentums
> This suggests that the limiting behavior is really a reduction in system momentum rather than prescribing an absolute limit

This also makes sense. `scroll-fling-decay` was attempting to be in line with this - don't give hard numbers, just hints to let the UA tweak it according to platform physics.

I suppose part of what I'm concerned with is making this feel true to other native scrollers on each platform. But perhaps that's more about UAs tweaking scroll code and less about customization.
